pypesto/objective/roadrunner/roadrunner_calculator.py

```diff
@@ -0,0 +1,403 @@
+            # obs_ss = steady_state[:, : len(observables_ids)].flatten()
+            state_ss = steady_state[:, len(observables_ids) :].flatten()
+            # turn off conserved moiety analysis
+            roadrunner_instance.conservedMoietyAnalysis = False
+            # reset the model
+            roadrunner_instance.reset()
+        # set parameters
+        par_map = self.fill_in_parameters(
+            x_dct, roadrunner_instance, parameter_mapping_per_condition
+        )
+        # if steady state calculations are required, set state variables
+        if steady_state_calculations:
+            roadrunner_instance.setValues(state_variables, state_ss)
+            # fill in overriden species
+            self.fill_in_parameters(
+                x_dct,
+                roadrunner_instance,
+                parameter_mapping_per_condition,
+                filling_mode="only_species",
+            )
+
+        sim_res = roadrunner_instance.simulate(
+            times=timepoints, selections=["time"] + observables_ids
+        )
+
+        llhs = calculate_llh(sim_res, edata, par_map)
+
+        # reset the model
+        roadrunner_instance.reset()
+
+        return sim_res, llhs
+
+    def fill_in_parameters(
+        self,
+        problem_parameters: dict,
+        roadrunner_instance: Optional[roadrunner.RoadRunner] = None,
+        parameter_mapping: Optional[ParMappingDictQuadruple] = None,
+        preeq: bool = False,
+        filling_mode: Optional[str] = None,
+    ) -> dict:
+        """Fill in parameters into the roadrunner instance.
+
+        Parameters
+        ----------
+        roadrunner_instance:
+            RoadRunner instance to fill in parameters
+        problem_parameters:
+            Problem parameters as parameterId=>value dict. Only
+            parameters included here will be set. Remaining parameters will
+            be used as already set in `amici_model` and `edata`.
+        parameter_mapping:
+            Parameter mapping for current condition.
+        preeq:
+            Whether to fill in parameters for pre-equilibration.
+        filling_mode:
+            Which parameters to fill in. If None or "all",
+            all parameters are filled in.
+            Other options are "only_parameters" and "only_species".
+
+        Returns
+        -------
+        dict:
+            Mapping of parameter IDs to values.
+        """
+        if filling_mode is None:
+            filling_mode = "all"
+        # check for valid filling modes
+        if filling_mode not in ["all", "only_parameters", "only_species"]:
+            raise ValueError(
+                "Invalid filling mode. Choose from 'all', "
+                "'only_parameters', 'only_species'."
+            )
+        mapping = parameter_mapping[1]
+        scaling = parameter_mapping[3]
+        if preeq:
+            mapping = parameter_mapping[0]
+            scaling = parameter_mapping[2]
```

Comment:
It is a bit tough to understand what the different entries of the parameter mapping are. Would it be better if `parameter_mapping` was a dictionary, with keys describing what kind of objects it contains? Similar to amici's `map_sim_var`, `map_sim_var_preeq` and so on.

Not crucial as this is really a under the hood kind of a thing, but would be easier to understand the code.

Reply:
added it into the doctring description and added comments. Think that should suffice for better code understanding for now.